prow/pod-utils/gcs/upload.go

```diff
@@ -0,0 +1,87 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gcs
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"sync"
+
+	"cloud.google.com/go/storage"
+	"github.com/sirupsen/logrus"
+)
+
+// UploadFunc knows how to upload into an object
+type UploadFunc func(obj *storage.ObjectHandle) error
+
+// UploadToGcs uploads all of the data in the
+// uploadTargets map to GCS in parallel. The map is
+// keyed on GCS path under the bucket
+func UploadToGcs(bucket *storage.BucketHandle, uploadTargets map[string]UploadFunc) error {
+	errCh := make(chan error, len(uploadTargets))
```

Comment:
Is it normal practice to use channels as thread safe arrays?

Seems like both this and the wait group is a bit redundant

Reply:
This is a practice seen even on the go blog: [https://blog.golang.org/pipelines#TOC_8.](https://blog.golang.org/pipelines#TOC_8.)

Reply:
I do not see the link you provided using a channel as an array.

The error channel has a fixed length of 1. The second file is not `case c <- result{}` until the first one is `for r := range c {}` As soon as an error is discovered the whole thing shuts down by returning out of `MD5All()` which will `close(done)`

Reply:
How else would you async aggregate errors from the goroutines?

Reply:
Here is one option
```
ec := make(chan error)
for d, u := range targets {
  go func(f, o) { ec <- f(o) }(u, o)
}
var errs []error
var n int
for n := 0; n < len(targets); n++ {
  if e <- ec; ec != nil {
    errs = append(errs, e)
  }
}
log.Printf("Found %d errors: %#v", len(errs), errs)
```